### DISCUSSION TITLE
Decoding dishwasher communication and some problems

### DISCUSSION THREAD

**Discussion by Battlecake91** [arendst/Tasmota#22502] [Q&A]
Hello,

I try to decode the D-Bus of my Siemens dishwasher. I connected a ESP32-C3 with a simple circuit to readout the D-Bus. 
Every device has a open drain connection to the bus and pulls down 5V. For my SN55M505EU/29 the communication is 9600 baud with 8N2. 

Now I'm able to see serial messages. 

First I checked the communication with a logic analyzer to know what I'm waiting for. It looks like this:
![2024-11-17 13_50_27-Window](https://github.com/user-attachments/assets/a248b703-931e-451f-bfc1-7bea9ffbc512)
![2024-11-17 13_52_36-Window](https://github.com/user-attachments/assets/8ce6a75b-406f-4f04-ad13-8346fa4db73b)

In Tasmota I get this: 
`13:48:11.834 RSL: RESULT = {"SSerialReceived":"05252008280000ADBE2A032520004C6A572A"}`
`13:49:11.754 RSL: RESULT = {"SSerialReceived":"05252008270000818F2A032520004D7A762A"}`
`13:50:11.592 RSL: RESULT = {"SSerialReceived":"05252008260000B6BF2A032520004E4A152A"}`

First I'm pretty happy to see valid data. But there are two messages together. It seems that they are stopped by "2A" or "1A" (I guess it depends what adress sends the end byte). My logic analyzer shows a pause of >10ms, but Tasmota seems to wait longer than that. 
Am I doing something wrong or is the timeout given by arduino or something else? I'm using software serial but can try it with hardware serial instead, if this could change something.

Anyway, I could try to decode it like it is and use the 2A or 1A as stop message, but it would be nice if the messages are correct separated.

Edit: The 2A/1A/0A isn't a "Stop byte", it doesn't comes from the same device so it seems to be a acknowledge from the receiver.

**Comment by joba-1** [comment]
Nice :) I think I know now why your name is Kampfkuchen. Subscribed.

Time between chars has no meaning on the level of the serial protocol. If at all, only on the d-bus level, but tasmota (or any serial driver) cannot and should not need to know this. 

If time doesn't really matter for the d-bus protocol you can concat the received messages and parse the resulting stream.
It looks like the first byte could be the data length and the remaining 3 bytes are header and/or checksum. If true it can work.

If you need the exact timing on ms level or even below you probably need to write your own C code. 
Which would be my prefered method during research anyways. Turnaround time for a small read serial, decode print loop would be much faster than rebuilding, linking and flashing the whole tasmota stuff on each change.

**Reply by Battlecake91** [reply]
Hah, thanks.

Yeah I thought I have to write it in C. I'm not fit with Berry (never used before) so C is a more pleasent for me. 
To get the information I want, it is not that important to have it separated directly, but I think this would be the more cleaner way. 

The goal should be to send messages, too. So better do the clean way from the beginning.
A friend wants to start the machine if the sun is shining long enough to power the dishwasher / washing machine with photovoltaics.

In this repo: [https://github.com/hn/bsh-home-appliances](url) there are really helpful information about the D-Bus. 
It's Length, Destination/Submodule, Data, Checksum in CRC16-XMODEM style and then the acknowledge answer. 

My SN55 has two free bus-connectors. One at the PCBA in the front and another on the main controller in the bottom. Both are parallel and free. 3 Pin with 2.54mm spacing. I'm searching for a good connector to just plug it in. At the moment I soldered some wires to the PCBA and connected it directly. 
When everything is ready and works I want design a PCB for a small ESP32 module with 3.3V buck and the D-Bus-electronics on it and a connector for a HLW8032 to measure the power consumption. 
The same works for my Siemens washing machine, it will get some, too. If I manage to get this working, I will share my design.

**Reply by joba-1** [reply]
That sounds pretty good! 

With such a pretty unintrusive low risk kind of edge connector/pinheader I might be tempted to try it out. Either to help testing if you need, or to just get rid of the cloud (I currently use the home assistant integration for home connect)

**Reply by Battlecake91** [reply]
My dishwasher is from around 2005, so no home connect. And my washing machine is from 2021, but fully integrated so I just had one option, unfortunately it was without home connect.
At the moment I measure the power and trigger events when the power goes under a given threshold.
But sometimes my dishwasher has a fault (like not able to pump) and my trigger doesn't work. Then after a hour or so I see, that it stopped. 
But please be careful, dishwasher and washing machines doesn't have a isolated power supply for the low voltages. It is possible to get shocked by the 5V, because to ground it has full line voltage. (I realized, that this depends on the orientation of the plug. In germany you can plug it in both orientations and in the worst case GND has 230V).

It would help me alot if you can support with the driver. I never wrote something for tasmota. So I have to do everything from scratch.

**Reply by joba-1** [reply]
I have written the tsl2561 driver but that was looong ago - not sure I would still recognize it (due to both, my memory and tasmota evolution). It became a bit pointless with all the solar panels in use today…
But yes I can support, still I think the usual suspects around here will be needed as well :)

Regarding the warning: thanks! That Gnd == 230V is non obvious, so I was about to write the same :)
Will only touch with WLAN or after unplugging and defusing the capacitors…and waiting

**Reply by Battlecake91** [reply]
Ah okay. No problem, I will try it by myself.

**Comment by joba-1** [comment]
May I ask how you connect to the d-bus? I also have a model named SN55*
